File: src/TeslaHub.Api/Utilities/PlusCode.cs
namespace TeslaHub.Api.Utilities;

/// <summary>
/// Encoder for Open Location Codes (a.k.a. "Plus Codes"), e.g.
/// <c>8FVC9G8F+6X</c>.
///
/// Why we ship our own encoder instead of `Google.OpenLocationCode`:
///   - One small, frozen-since-2014 algorithm; no transitive
///     dependency required.
///   - Keeps the file count for the Tesla integration self-contained.
///
/// Why TeslaHub uses Plus Codes for "Send to car":
///   - Tesla's <c>command/share</c> parser geocodes any text it
///     receives. Raw <c>lat,lng</c>, <c>"Address\nlat,lng"</c> and even
///     <c>https://maps.google.com/?q=lat,lng</c> URLs all funnel through
///     a fuzzy address resolver and snap the displayed destination to
///     the closest known POI/road, drifting tens to hundreds of meters
///     off the actual pin.
///   - Plus Codes are parsed as a pure geographic cell — there is no
///     address resolution, no POI fuzzing. The cell IS the destination.
///   - Cross-verified by the Tesla owner community as the most reliable
///     way to drop a destination on an exact spot (parking spaces,
///     trailheads, off-road pins, …).
///
/// Length / precision (per the OLC spec):
///   - 10 chars (PAIR_CODE_LENGTH)               → ~14 m × 14 m cell
///   - 11 chars (recommended for Tesla nav)      → ~2.8 m × 3.5 m cell
///   - 12 chars                                  → ~0.7 m × 0.6 m cell
///
/// Algorithm reference:
///   https://github.com/google/open-location-code/blob/main/docs/specification.md
///   https://github.com/google/open-location-code/blob/main/js/src/openlocationcode.js
///   (Apache-2.0, ported faithfully).
/// </summary>
public static class PlusCode
{
    private const string CodeAlphabet = "23456789CFGHJMPQRVWX";
    private const int EncodingBase = 20;
    private const int LatitudeMax = 90;
    private const int LongitudeMax = 180;
    private const int MaxDigitCount = 15;
    private const int PairCodeLength = 10;
    private const int GridCodeLength = MaxDigitCount - PairCodeLength; // 5
    private const int GridColumns = 4;
    private const int GridRows = 5;
    private const char Separator = '+';
    private const int SeparatorPosition = 8;

    // PAIR_PRECISION = ENCODING_BASE^3 = 8000. NOT EncodingBase^5 — the
    // two most-significant pair digits encode 20° / 1° per cell and live
    // OUTSIDE this multiplier. Match the canonical reference exactly,
    // misreading this constant produces codes that are ~400× too coarse.
    private const long PairPrecision = 8000L; // 20^3
    private const long FinalLatPrecision = PairPrecision * 3125L; // * 5^5
    private const long FinalLngPrecision = PairPrecision * 1024L; // * 4^5

    /// <summary>
    /// Encodes the given latitude/longitude into a Plus Code. The default
    /// length of 11 characters yields ~3 m precision — the sweet spot
    /// for Tesla in-vehicle navigation (matches civilian GPS accuracy
    /// without wasting bits).
    /// </summary>
    public static string Encode(double latitude, double longitude, int codeLength = 11)
    {
        if (codeLength < 2 || (codeLength < PairCodeLength && codeLength % 2 == 1))
            throw new ArgumentOutOfRangeException(nameof(codeLength),
                $"Invalid Open Location Code length: {codeLength}");

        var length = Math.Min(codeLength, MaxDigitCount);

        var lat = Math.Max(-LatitudeMax, Math.Min(LatitudeMax, latitude));
        var lng = longitude;
        while (lng < -LongitudeMax) lng += 360;
        while (lng >= LongitudeMax) lng -= 360;

        // Move into a positive integer space to keep all subsequent
        // arithmetic floating-point safe (mirrors the canonical lib).
        long latVal = (long)Math.Floor(lat * FinalLatPrecision) + (long)LatitudeMax * FinalLatPrecision;
        if (latVal < 0) latVal = 0;
        if (latVal >= 2L * LatitudeMax * FinalLatPrecision)
            latVal = 2L * LatitudeMax * FinalLatPrecision - 1;

        long lngVal = (long)Math.Floor(lng * FinalLngPrecision) + (long)LongitudeMax * FinalLngPrecision;
        var lngRange = 2L * LongitudeMax * FinalLngPrecision;
        if (lngVal < 0)
            lngVal = (lngVal % lngRange) + lngRange;
        else if (lngVal >= lngRange)
            lngVal %= lngRange;

        var code = string.Empty;

        if (length > PairCodeLength)
        {
            // Grid section — 4×5 sub-cells per refinement step. Each
            // additional grid character shrinks the cell area by 20×.
            for (var i = 0; i < GridCodeLength; i++)
            {
                var latDigit = (int)(latVal % GridRows);
                var lngDigit = (int)(lngVal % GridColumns);
                code = CodeAlphabet[latDigit * GridColumns + lngDigit] + code;
                latVal /= GridRows;
                lngVal /= GridColumns;
            }
        }
        else
        {
            latVal /= (long)Math.Pow(GridRows, GridCodeLength);
            lngVal /= (long)Math.Pow(GridColumns, GridCodeLength);
        }

        // Pair section — 5 (lat, lng) pairs = 10 digits, separator
        // inserted right after the first 8 digits (XXXXXXXX+XX layout).
        for (var i = 0; i < PairCodeLength / 2; i++)
        {
            code = CodeAlphabet[(int)(lngVal % EncodingBase)] + code;
            code = CodeAlphabet[(int)(latVal % EncodingBase)] + code;
            latVal /= EncodingBase;
            lngVal /= EncodingBase;
            if (i == 0) code = Separator + code;
        }

        if (length >= SeparatorPosition)
            return code[..(length + 1)];

        return code[..length] + new string('0', SeparatorPosition - length) + Separator;
    }
}

File: src/TeslaHub.Web/src/utils/plusCode.ts
/**
 * Encodes a (latitude, longitude) pair into an Open Location Code
 * (a.k.a. "Plus Code", e.g. `8FVC9G8F+6X`).
 *
 * Why we ship our own encoder instead of `npm i open-location-code`:
 *   - One small algorithm, frozen since 2014, no upstream churn.
 *   - Avoids pulling a transitive dependency and CommonJS/ESM friction
 *     for a ~3 KB function.
 *
 * Why we use Plus Codes for the Tesla "Send to car" feature:
 *   - Tesla's `command/share` parser geocodes any text it receives. If
 *     we hand it raw `lat,lng`, the in-vehicle navigation snaps to the
 *     nearest known address/POI/road segment, often dropping the
 *     destination tens to hundreds of meters off the actual map pin.
 *   - Google Maps URLs (`?q=lat,lng`) help (Tesla recognises them) but
 *     the routing engine still snaps to the closest road for routing,
 *     and the displayed pin can drift if the area has a strong POI.
 *   - Plus Codes are parsed by Tesla as a pure geographic cell, NOT as
 *     a search query. The cell IS the destination, so there is nothing
 *     to fuzzy-match. The community has cross-verified this works for
 *     parking spots, off-road pins, and general meter-precision use.
 *
 * Length / precision (per the OLC spec):
 *   - 10 chars (PAIR_CODE_LENGTH)               → ~14 m × 14 m cell
 *   - 11 chars (recommended default for Tesla)  → ~2.8 m × 3.5 m cell
 *   - 12 chars                                  → ~0.7 m × 0.6 m cell
 *   The recommended 11-char form matches consumer GPS accuracy
 *   (~3 m) so going higher is wasted bits.
 *
 * Algorithm reference:
 *   https://github.com/google/open-location-code/blob/main/docs/specification.md
 *   https://github.com/google/open-location-code/blob/main/js/src/openlocationcode.js
 *   (Apache-2.0)
 */

const CODE_ALPHABET = '23456789CFGHJMPQRVWX';
const ENCODING_BASE = 20;
const LATITUDE_MAX = 90;
const LONGITUDE_MAX = 180;
const MAX_DIGIT_COUNT = 15;
const PAIR_CODE_LENGTH = 10;
const GRID_CODE_LENGTH = MAX_DIGIT_COUNT - PAIR_CODE_LENGTH; // 5
const GRID_COLUMNS = 4;
const GRID_ROWS = 5;
// Multiplier that converts a raw degree value into the pair-section
// integer space (covers up to ~0.000125° per cell at finest pair step).
// Per the OLC reference implementation, this is ENCODING_BASE ** 3 = 8000
// — NOT ENCODING_BASE ** 5. The 2 most-significant pair digits encode
// 20° and 1° per cell respectively and they live OUTSIDE this multiplier.
const PAIR_PRECISION = Math.pow(ENCODING_BASE, 3);
const FINAL_LAT_PRECISION = PAIR_PRECISION * Math.pow(GRID_ROWS, GRID_CODE_LENGTH);
const FINAL_LNG_PRECISION = PAIR_PRECISION * Math.pow(GRID_COLUMNS, GRID_CODE_LENGTH);
const SEPARATOR = '+';
const SEPARATOR_POSITION = 8;

/**
 * Returns the Open Location Code (Plus Code) for the given coordinates.
 * `codeLength` defaults to 11 — the sweet spot for Tesla navigation:
 * ~3 m precision (matching civilian GPS) without unnecessary length.
 */
export function encodePlusCode(
  latitude: number,
  longitude: number,
  codeLength: number = 11,
): string {
  if (codeLength < 2 || (codeLength < PAIR_CODE_LENGTH && codeLength % 2 === 1)) {
    throw new Error(`Invalid Open Location Code length: ${codeLength}`);
  }
  const length = Math.min(codeLength, MAX_DIGIT_COUNT);

  let lat = Math.max(-LATITUDE_MAX, Math.min(LATITUDE_MAX, latitude));
  let lng = longitude;
  while (lng < -LONGITUDE_MAX) lng += 360;
  while (lng >= LONGITUDE_MAX) lng -= 360;

  // Move into a positive integer space to keep all subsequent maths
  // floating-point safe (the same trick the canonical JS library uses).
  let latVal = Math.floor(lat * FINAL_LAT_PRECISION) + LATITUDE_MAX * FINAL_LAT_PRECISION;
  if (latVal < 0) latVal = 0;
  if (latVal >= 2 * LATITUDE_MAX * FINAL_LAT_PRECISION) {
    latVal = 2 * LATITUDE_MAX * FINAL_LAT_PRECISION - 1;
  }
  let lngVal = Math.floor(lng * FINAL_LNG_PRECISION) + LONGITUDE_MAX * FINAL_LNG_PRECISION;
  if (lngVal < 0) {
    lngVal = (lngVal % (2 * LONGITUDE_MAX * FINAL_LNG_PRECISION)) +
      2 * LONGITUDE_MAX * FINAL_LNG_PRECISION;
  } else if (lngVal >= 2 * LONGITUDE_MAX * FINAL_LNG_PRECISION) {
    lngVal = lngVal % (2 * LONGITUDE_MAX * FINAL_LNG_PRECISION);
  }

  let code = '';

  if (length > PAIR_CODE_LENGTH) {
    // Grid section — 4×5 sub-cells per refinement step, each character
    // contributes one extra grid digit so each step shrinks the cell by
    // a factor of 20 in area.
    for (let i = 0; i < GRID_CODE_LENGTH; i++) {
      const latDigit = latVal % GRID_ROWS;
      const lngDigit = lngVal % GRID_COLUMNS;
      code = CODE_ALPHABET.charAt(latDigit * GRID_COLUMNS + lngDigit) + code;
      latVal = Math.floor(latVal / GRID_ROWS);
      lngVal = Math.floor(lngVal / GRID_COLUMNS);
    }
  } else {
    // Drop the grid bits when the caller only wanted pair-precision.
    latVal = Math.floor(latVal / Math.pow(GRID_ROWS, GRID_CODE_LENGTH));
    lngVal = Math.floor(lngVal / Math.pow(GRID_COLUMNS, GRID_CODE_LENGTH));
  }

  // Pair section — 5 (lat, lng) pairs = 10 digits, separator goes in
  // after the first 8 digits so the output is "XXXXXXXX+XX" style.
  for (let i = 0; i < PAIR_CODE_LENGTH / 2; i++) {
    code = CODE_ALPHABET.charAt(lngVal % ENCODING_BASE) + code;
    code = CODE_ALPHABET.charAt(latVal % ENCODING_BASE) + code;
    latVal = Math.floor(latVal / ENCODING_BASE);
    lngVal = Math.floor(lngVal / ENCODING_BASE);
    if (i === 0) {
      code = SEPARATOR + code;
    }
  }

  // The full code is 16 chars (15 digits + 1 separator); trim back to
  // the requested length, keeping the separator intact.
  if (length >= SEPARATOR_POSITION) {
    return code.substring(0, length + 1);
  }
  // Padding-character branch is unreachable here because we always ask
  // for length >= SEPARATOR_POSITION (>= 8), but kept for completeness
  // matching the reference implementation.
  return (
    code.substring(0, length) +
    new Array(SEPARATOR_POSITION - length + 1).join('0') +
    SEPARATOR
  );
}
